### PR TITLE
Add CP subsytem config for Hazelcast CR [CN-1149]

### PR DESCRIPTION
**Pull request by SeriyBg**
## User Impact

Add the possibility to configure CP Subsystem in the Hazelcast CR

### CODE REVIEW

File: api/v1alpha1/hazelcast_validation.go [hunk 1]
```diff
@@ -427,3 +428,28 @@ func (v *hazelcastValidator) validateSQL(h *Hazelcast) {
...
+
+func (v *hazelcastValidator) validateCPSubsystem(h *Hazelcast) {
+	if h.Spec.CPSubsystem == nil {
+		return
+	}
+
+	cp := h.Spec.CPSubsystem
+	if h.Spec.ClusterSize == nil {
+		if cp.MemberCount > 3 {
+			v.Invalid(Path("spec", "cpSubsystem", "memberCount"), cp.MemberCount, "can not be greater the clusterSize")
```
Comment (cagric0) [inline api/v1alpha1/hazelcast_validation.go:440]: Shouldn't we use different error messages for different cases to distinguish the failure reason?

File: api/v1alpha1/hazelcast_validation.go [hunk 2]
```diff
@@ -427,3 +428,28 @@ func (v *hazelcastValidator) validateSQL(h *Hazelcast) {
...
+	}
+
+	if cp.GroupSize != nil {
+		if (*cp.GroupSize != 3 && *cp.GroupSize != 5 && *cp.GroupSize != 7) || *cp.GroupSize > cp.MemberCount {
+			v.Invalid(Path("spec", "cpSubsystem", "memberCount"), cp.GroupSize, "can be 3, 5, or 7, but not greater that memberCount")
+		}
+	}
+
+	if cp.PVC == nil && h.Spec.Persistence.Pvc == nil {
+		v.Required(Path("spec", "cpSubsystem", "pvc"), "PVC should be configured")
```
Comment (cagric0) [inline api/v1alpha1/hazelcast_validation.go:453]: The point is if cp is enabled either cp.Pvc or persistence.Pvc must be configured?

File: api/v1alpha1/hazelcast_validation.go [hunk 3]
```diff
@@ -427,3 +428,28 @@ func (v *hazelcastValidator) validateSQL(h *Hazelcast) {
...
+		}
+	} else if cp.MemberCount > *h.Spec.ClusterSize && *h.Spec.ClusterSize != 0 { // ClusterSize of 0 means pausing the cluster, should be ignored
+		v.Invalid(Path("spec", "cpSubsystem", "memberCount"), cp.MemberCount, "can not be greater the clusterSize")
+	}
+
+	if cp.GroupSize != nil {
+		if (*cp.GroupSize != 3 && *cp.GroupSize != 5 && *cp.GroupSize != 7) || *cp.GroupSize > cp.MemberCount {
+			v.Invalid(Path("spec", "cpSubsystem", "memberCount"), cp.GroupSize, "can be 3, 5, or 7, but not greater that memberCount")
+		}
+	}
```
Comment (cagric0) [inline api/v1alpha1/hazelcast_validation.go:450]: Can `GroupSize` be less than `MemberCount`?

File: api/v1alpha1/hazelcast_validation.go [hunk 4]
```diff
@@ -427,3 +428,28 @@ func (v *hazelcastValidator) validateSQL(h *Hazelcast) {
 		v.Forbidden(Path("spec", "sql", "catalogPersistence"), "catalogPersistence requires Hazelcast persistence enabled")
 	}
 }
+
+func (v *hazelcastValidator) validateCPSubsystem(h *Hazelcast) {
```
Comment (cagric0) [inline api/v1alpha1/hazelcast_validation.go:432]: We are missing the integration tests for these validations. Could you add them?

**Review by cagric0 — CHANGES_REQUESTED** [review]

File: api/v1alpha1/hazelcast_validation.go (hunk 1 above)
Comment (SeriyBg) [reply api/v1alpha1/hazelcast_validation.go:440]: But this is the same reason. The only thing is that the `ClusterSize` is not specified, then we use the default `3` member count

File: api/v1alpha1/hazelcast_validation.go (hunk 3 above)
Comment (SeriyBg) [reply api/v1alpha1/hazelcast_validation.go:450]: Yes, it can be less but not greater than

**Review by cagric0 — APPROVED** [review]

**Pull request merged**